pkg/labels/selector.go

```diff
@@ -303,24 +692,19 @@ func (lsel *LabelSelector) String() (string, error) {
 // as they parse different selectors with different syntaxes.
 // The input will cause an error if it does not follow this form:
 //
-//     <selector-syntax> ::= <requirement> | <requirement> "," <selector-syntax>
-//         <requirement> ::= WHITESPACE_OPT KEY <set-restriction>
-//     <set-restriction> ::= "" | <inclusion-exclusion> <value-set>
+// <selector-syntax> ::= <requirement> | <requirement> "," <selector-syntax> ]
+// <requirement> ::= KEY [ <set-based-restriction> | <exact-match-restriction>
+// <set-based-restriction> ::= "" | <inclusion-exclusion> <value-set>
 // <inclusion-exclusion> ::= <inclusion> | <exclusion>
-//           <exclusion> ::= WHITESPACE "not" <inclusion>
-//           <inclusion> ::= WHITESPACE "in" WHITESPACE
+//           <exclusion> ::= "not" <inclusion>
+//           <inclusion> ::= "in"
 //           <value-set> ::= "(" <values> ")"
 //              <values> ::= VALUE | VALUE "," <values>
-//
-// KEY is a sequence of one or more characters that does not contain ',' or ' '
-//      [^, ]+
-// VALUE is a sequence of zero or more characters that does not contain ',', ' ' or ')'
-//      [^, )]*
-// WHITESPACE_OPT is a sequence of zero or more whitespace characters
-//      \s*
-// WHITESPACE is a sequence of one or more whitespace characters
-//      \s+
-//
+// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
+// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
+// VALUE is a sequence of zero or more characters that does not contain: ('=', '!', '(', ')', ',')
```

Comment:
I can't find it now - maybe I'm remembering the last face to face where we talked about being strict on labels to make the UI easier and clearer to read.

Using the value of a label as an input to machine (for processing) seems to be what annotations are for.  Wanting to find things by arbitrary annotations seems like an argument for exposing search/query on annotations via some extra service.  Wanting to find things by clear, unambiguous groupings and arrangements is labels, and it's intended to bridge a gap between machines (which work on the groupings) and humans (which need to understand the groupings).  Long, complex label values will be hard to use in UIs and from the CLI (certainly anything require bash escaping, or involving quotes, is doomed).  Having spaces in label values (unless clearly defined now) does seem like it opens the doors to "label query injection attacks" in the future.  The options are LabelPreparedStatements, or a restrictive value syntax.  If labels are like tagging systems, then there's reasonable practical experience saying spaces don't make or break.

Those are the thoughts in my head right now.

----- Original Message -----
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ','
> > or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain
> > ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> 
> Which arguments were these?  Last I saw there was a hangup about labels and
> annotations being different, but I am increasingly convince they ARE
> different.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24871642